File: packages/typescript/src/http.ts
/**
 * HTTP transport: the single place that talks to the network. Owns
 * authentication, JSON encoding, timeouts, retries with backoff, and turning
 * non-2xx responses into {@link AxeneError}. Resources depend on this, not on
 * `fetch` directly.
 * @module
 */
import { AxeneError } from './errors';
import type { AxeneOptions } from './types';

const DEFAULT_BASE = 'https://mail.axene.io';

const sleep = (ms: number) => new Promise((resolve) => setTimeout(resolve, ms));

/** Resolved, validated transport configuration. */
type ResolvedConfig = Required<Omit<AxeneOptions, 'fetch'>> & { fetch: typeof fetch };

export class HttpTransport {
  private readonly cfg: ResolvedConfig;

  constructor(options: AxeneOptions) {
    if (!options?.apiKey) throw new Error('Axene: `apiKey` is required.');
    const f = options.fetch ?? globalThis.fetch;
    if (!f) throw new Error('Axene: no global fetch found. Pass `fetch` in options (Node < 18).');
    this.cfg = {
      apiKey: options.apiKey,
      baseUrl: (options.baseUrl ?? DEFAULT_BASE).replace(/\/+$/, ''),
      maxRetries: options.maxRetries ?? 3,
      timeoutMs: options.timeoutMs ?? 30000,
      fetch: f,
    };
  }

  /**
   * Perform a request and parse the JSON response.
   *
   * Retries `429` and `5xx` with exponential backoff (honoring `Retry-After`
   * when present). Throws {@link AxeneError} on a final non-2xx or a transport
   * failure that survives all attempts.
   */
  async request<T>(method: string, path: string, body?: unknown): Promise<T> {
    const url = `${this.cfg.baseUrl}${path}`;
    let lastError: unknown;

    for (let attempt = 1; attempt <= this.cfg.maxRetries; attempt++) {
      const controller = new AbortController();
      const timer = setTimeout(() => controller.abort(), this.cfg.timeoutMs);
      try {
        const res = await this.cfg.fetch(url, {
          method,
          headers: {
            Authorization: `Bearer ${this.cfg.apiKey}`,
            'Content-Type': 'application/json',
            'User-Agent': '@axene/mailer',
          },
          body: body === undefined ? undefined : JSON.stringify(body),
          signal: controller.signal,
        });
        clearTimeout(timer);

        if (this.isRetryable(res.status) && attempt < this.cfg.maxRetries) {
          await sleep(this.backoffMs(res, attempt));
          continue;
        }

        const payload = await this.parseBody(res);
        if (!res.ok) throw this.toError(res.status, payload);
        return payload as T;
      } catch (err) {
        clearTimeout(timer);
        if (err instanceof AxeneError) throw err; // a real API error: do not retry
        lastError = err; // transport/abort error: retry if attempts remain
        if (attempt < this.cfg.maxRetries) {
          await sleep(this.backoffMs(undefined, attempt));
          continue;
        }
      }
    }
    throw new AxeneError(0, `Axene request failed: ${String(lastError)}`);
  }

  /**
   * Upload a single file as `multipart/form-data` under the field name `file`.
   * Used by the CSV import endpoints. Does not set `Content-Type` so the runtime
   * adds the correct multipart boundary. Not retried (uploads are not idempotent).
   */
  async upload<T>(path: string, file: Uint8Array, filename: string): Promise<T> {
    const url = `${this.cfg.baseUrl}${path}`;
    const form = new FormData();
    const blob = new Blob([file.slice()]);
    form.append('file', blob, filename);

    const controller = new AbortController();
    const timer = setTimeout(() => controller.abort(), this.cfg.timeoutMs);
    try {
      const res = await this.cfg.fetch(url, {
        method: 'POST',
        headers: { Authorization: `Bearer ${this.cfg.apiKey}`, 'User-Agent': '@axene/mailer' },
        body: form as unknown as BodyInit,
        signal: controller.signal,
      });
      const payload = await this.parseBody(res);
      if (!res.ok) throw this.toError(res.status, payload);
      return payload as T;
    } finally {
      clearTimeout(timer);
    }
  }

  private isRetryable(status: number): boolean {
    return status === 429 || status >= 500;
  }

  private backoffMs(res: Response | undefined, attempt: number): number {
    const retryAfter = res ? Number(res.headers.get('retry-after')) : NaN;
    if (Number.isFinite(retryAfter) && retryAfter > 0) return retryAfter * 1000;
    return 250 * 2 ** (attempt - 1);
  }

  private async parseBody(res: Response): Promise<unknown> {
    const isJson = (res.headers.get('content-type') ?? '').includes('application/json');
    return isJson ? res.json().catch(() => undefined) : undefined;
  }

  /** Map the API's `{ detail: { code, message } }` (or string) into an {@link AxeneError}. */
  private toError(status: number, payload: unknown): AxeneError {
    const detail = (payload as { detail?: unknown } | undefined)?.detail;
    const code = typeof detail === 'object' && detail ? (detail as { code?: string }).code : undefined;
    const message =
      (typeof detail === 'object' && detail ? (detail as { message?: string }).message : undefined) ??
      (typeof detail === 'string' ? detail : undefined) ??
      `Axene request failed (${status})`;
    return new AxeneError(status, message, code, payload);
  }
}

File: packages/rust/src/http.rs
//! HTTP transport: the single place that talks to the network.
//!
//! Owns authentication (bearer), JSON encode/decode, retries with backoff on
//! `429`/`5xx` (honouring `Retry-After`), multipart upload, and turning non-2xx
//! responses into [`AxeneError`]. Resources depend on this, not on `reqwest`
//! directly.

use std::time::Duration;

use reqwest::{multipart, Client, Method, Response, StatusCode};
use serde::de::DeserializeOwned;
use serde::Serialize;
use serde_json::Value;

use crate::error::{AxeneError, Result};

const DEFAULT_BASE: &str = "https://mail.axene.io";
const USER_AGENT: &str = concat!("axene-mailer-rust/", env!("CARGO_PKG_VERSION"));

/// The network transport shared by every resource.
#[derive(Debug, Clone)]
pub struct HttpTransport {
    client: Client,
    api_key: String,
    base_url: String,
    max_retries: u32,
}

impl HttpTransport {
    /// Build a transport. `base_url` has any trailing slashes trimmed.
    pub(crate) fn new(
        api_key: String,
        base_url: Option<String>,
        max_retries: u32,
        timeout: Duration,
    ) -> Result<Self> {
        if api_key.is_empty() {
            return Err(AxeneError::transport("Axene: `api_key` is required."));
        }
        let base = base_url.unwrap_or_else(|| DEFAULT_BASE.to_string());
        let base = base.trim_end_matches('/').to_string();
        let client = Client::builder()
            .timeout(timeout)
            .build()
            .map_err(|e| AxeneError::transport(format!("failed to build HTTP client: {e}")))?;
        Ok(Self {
            client,
            api_key,
            base_url: base,
            // At least one attempt.
            max_retries: max_retries.max(1),
        })
    }

    /// Send a request with no body and decode the JSON response.
    pub(crate) async fn request<T: DeserializeOwned>(
        &self,
        method: Method,
        path: &str,
    ) -> Result<T> {
        self.send::<(), T>(method, path, None).await
    }

    /// Send a request with a JSON body and decode the JSON response.
    pub(crate) async fn request_json<B: Serialize, T: DeserializeOwned>(
        &self,
        method: Method,
        path: &str,
        body: &B,
    ) -> Result<T> {
        self.send(method, path, Some(body)).await
    }

    /// Send a request that returns no content (204 / empty body).
    pub(crate) async fn request_empty(&self, method: Method, path: &str) -> Result<()> {
        let _: IgnoredBody = self.send::<(), IgnoredBody>(method, path, None).await?;
        Ok(())
    }

    async fn send<B: Serialize, T: DeserializeOwned>(
        &self,
        method: Method,
        path: &str,
        body: Option<&B>,
    ) -> Result<T> {
        let url = format!("{}{}", self.base_url, path);
        let mut last_error: Option<AxeneError> = None;

        for attempt in 1..=self.max_retries {
            let mut builder = self
                .client
                .request(method.clone(), &url)
                .bearer_auth(&self.api_key)
                .header(reqwest::header::USER_AGENT, USER_AGENT);
            if let Some(b) = body {
                builder = builder.json(b);
            }

            match builder.send().await {
                Ok(res) => {
                    let status = res.status();
                    if is_retryable(status) && attempt < self.max_retries {
                        let wait = backoff(Some(&res), attempt);
                        last_error = Some(error_placeholder(status));
                        tokio::time::sleep(wait).await;
                        continue;
                    }
                    return decode(res).await;
                }
                Err(err) => {
                    last_error = Some(AxeneError::transport(format!(
                        "Axene request failed: {err}"
                    )));
                    if attempt < self.max_retries {
                        tokio::time::sleep(backoff(None, attempt)).await;
                        continue;
                    }
                }
            }
        }

        Err(last_error
            .unwrap_or_else(|| AxeneError::transport("Axene request failed: no attempts made")))
    }

    /// Upload a single file as `multipart/form-data` under the field `file`.
    ///
    /// Not retried (uploads are not idempotent). The runtime sets the multipart
    /// boundary; we do not set `Content-Type` manually.
    pub(crate) async fn upload<T: DeserializeOwned>(
        &self,
        path: &str,
        file: Vec<u8>,
        filename: &str,
    ) -> Result<T> {
        let url = format!("{}{}", self.base_url, path);
        let part = multipart::Part::bytes(file).file_name(filename.to_string());
        let form = multipart::Form::new().part("file", part);

        let res = self
            .client
            .post(&url)
            .bearer_auth(&self.api_key)
            .header(reqwest::header::USER_AGENT, USER_AGENT)
            .multipart(form)
            .send()
            .await
            .map_err(|e| AxeneError::transport(format!("Axene upload failed: {e}")))?;
        decode(res).await
    }
}

/// A unit-like type that deserializes from any (or empty) body.
#[derive(serde::Deserialize)]
struct IgnoredBody {
    #[serde(flatten)]
    #[allow(dead_code)]
    rest: std::collections::HashMap<String, Value>,
}

fn is_retryable(status: StatusCode) -> bool {
    status == StatusCode::TOO_MANY_REQUESTS || status.is_server_error()
}

fn backoff(res: Option<&Response>, attempt: u32) -> Duration {
    if let Some(res) = res {
        if let Some(secs) = res
            .headers()
            .get(reqwest::header::RETRY_AFTER)
            .and_then(|v| v.to_str().ok())
            .and_then(|v| v.trim().parse::<f64>().ok())
        {
            if secs > 0.0 {
                return Duration::from_secs_f64(secs);
            }
        }
    }
    // 250ms, 500ms, 1s, ...
    Duration::from_millis(250 * 2u64.pow(attempt.saturating_sub(1)))
}

/// Used only to record the last status when a retry loop exhausts.
fn error_placeholder(status: StatusCode) -> AxeneError {
    AxeneError::new(
        status.as_u16(),
        format!("Axene request failed ({})", status.as_u16()),
        None,
    )
}

/// Decode a response: success bodies parse into `T`, errors map to [`AxeneError`].
async fn decode<T: DeserializeOwned>(res: Response) -> Result<T> {
    let status = res.status();
    let bytes = res
        .bytes()
        .await
        .map_err(|e| AxeneError::transport(format!("failed to read response body: {e}")))?;

    if status.is_success() {
        if bytes.is_empty() {
            // 204 No Content, or an empty body for a `()`/unit decode.
            return serde_json::from_slice(b"{}").map_err(|e| {
                AxeneError::transport(format!("failed to decode empty body: {e}"))
            });
        }
        return serde_json::from_slice(&bytes).map_err(|e| {
            AxeneError::transport(format!("failed to decode response: {e}"))
        });
    }

    Err(to_error(status.as_u16(), &bytes))
}

/// Map the API's `{ detail: { code, message } }` (or string) into an error.
fn to_error(status: u16, bytes: &[u8]) -> AxeneError {
    let payload: Option<Value> = serde_json::from_slice(bytes).ok();
    let detail = payload.as_ref().and_then(|p| p.get("detail"));

    let mut code = None;
    let mut message = None;

    if let Some(detail) = detail {
        if let Some(obj) = detail.as_object() {
            code = obj.get("code").and_then(|c| c.as_str()).map(String::from);
            message = obj.get("message").and_then(|m| m.as_str()).map(String::from);
        } else if let Some(s) = detail.as_str() {
            message = Some(s.to_string());
        }
    }

    AxeneError::new(
        status,
        message.unwrap_or_else(|| format!("Axene request failed ({status})")),
        code,
    )
}
